planner/core/task.go

```diff
@@ -1578,6 +1578,35 @@ func (p *basePhysicalAgg) newPartialAggregate(copTaskType kv.StoreType, isMPPTas
 	return partialAgg, finalAgg
 }
 
+// canUse3StageDistinctAgg returns true if this agg can use 3 stage for distinct aggregation
+func (p *basePhysicalAgg) canUse3StageDistinctAgg() bool {
+	num := 0
+	if !p.ctx.GetSessionVars().Enable3StageDistinctAgg || len(p.GroupByItems) > 0 {
+		return false
+	}
+	for _, fun := range p.AggFuncs {
+		if fun.HasDistinct {
+			num++
+			if num > 1 || fun.Name != ast.AggFuncCount {
+				return false
+			}
+			for _, arg := range fun.Args {
+				// bail out when args are not simple column, see GitHub issue #35417
+				if _, ok := arg.(*expression.Column); !ok {
+					return false
+				}
+			}
+		} else if len(fun.Args) > 1 {
+			return false
```

Comment:
why limit this here？any cases about it？

Reply:
`JSON_OBJECTAGG`, or stats agg function `corr(Y, X)` (not supported yet), have multiple args. Do not bother splitting them for the moment. `JSON_OBJECTAGG` is not supported by TiFlash yet, but still added test cases.